execution/block-partitioner/src/sharded_block_partitioner/partitioning_shard.rs

```diff
@@ -0,0 +1,242 @@
+            }
+        }
+        rw_set_with_index_vec
+    }
+
+    fn broadcast_num_accepted_txns(&self, num_accepted_txns: usize) {
+        let num_shards = self.message_txs.len();
+        for i in 0..num_shards {
+            if i != self.shard_id {
+                self.message_txs[i]
+                    .send(CrossShardMsg::AcceptedTxnsMsg(num_accepted_txns))
+                    .unwrap();
+            }
+        }
+    }
+
+    fn collect_num_accepted_txns(&self) -> Vec<usize> {
+        let mut accepted_txns_vec = vec![0; self.message_txs.len()];
+        for (i, msg_rx) in self.message_rxs.iter().enumerate() {
+            if i == self.shard_id {
+                continue;
+            }
+            let msg = msg_rx.recv().unwrap();
+            match msg {
+                CrossShardMsg::AcceptedTxnsMsg(num_accepted_txns) => {
+                    accepted_txns_vec[i] = num_accepted_txns;
+                },
+                _ => panic!("Unexpected message"),
+            }
+        }
+        accepted_txns_vec
+    }
+
+    fn discard_txns_with_cross_shard_deps(&self, partition_msg: DiscardTxnsWithCrossShardDep) {
+        let DiscardTxnsWithCrossShardDep {
+            transactions,
+            prev_rounds_rw_set_with_index,
+            prev_rounds_frozen_chunks,
+        } = partition_msg;
+        let num_shards = self.message_txs.len();
+        let mut conflict_detector = CrossShardConflictDetector::new(self.shard_id, num_shards);
+        // If transaction filtering is allowed, we need to prepare the dependency analysis and broadcast it to other shards
+        // Based on the dependency analysis received from other shards, we will reject transactions that are conflicting with
+        // transactions in other shards
+        let read_write_set = RWSet::new(&transactions);
+        self.broadcast_rw_set(read_write_set);
+        let cross_shard_rw_set = self.collect_rw_set();
+        let (accepted_txns, accepted_cross_shard_dependencies, rejected_txns) = conflict_detector
+            .discard_txns_with_cross_shard_deps(
+                transactions,
+                &cross_shard_rw_set,
+                prev_rounds_rw_set_with_index,
+            );
+        // Broadcast and collect the stats around number of accepted and rejected transactions from other shards
+        // this will be used to determine the absolute index of accepted transactions in this shard.
+        self.broadcast_num_accepted_txns(accepted_txns.len());
+        let accepted_txns_vec = self.collect_num_accepted_txns();
+        // Calculate the absolute index of accepted transactions in this shard, which is the sum of all accepted transactions
+        // from other shards whose shard id is smaller than the current shard id and the number of accepted transactions in the
+        // previous rounds
+        let mut index_offset = prev_rounds_frozen_chunks
+            .iter()
+            .map(|chunk| chunk.len())
+            .sum::<usize>();
```

Comment:
it seems wasteful to do `sum()` repeatedly, how about tracking it with a number across rounds?

Reply:
It doesn't add too much overhead as the number of chunks are not that hight, doing this in the master actually might make it more expensive because of single thread computation. Let me leave a TODO here to re-evaluate if it becomes expensive?

Reply:
That's fine.